### PR TITLE
WIP: Prepare Thanos for stringlabels

### PR DESCRIPTION
**Pull request by MichaHoffmann**

### CODE REVIEW

File: pkg/store/labelpb/label.go [hunk 1]
```diff
@@ -292,30 +292,29 @@ func (m *ZLabel) Compare(other ZLabel) int {
...
-			ret = append(ret, extend[0])
-			lset, extend = lset[1:], extend[1:]
-		} else if d < 0 {
-			ret = append(ret, lset[0])
-			lset = lset[1:]
-		} else if d > 0 {
-			ret = append(ret, extend[0])
-			extend = extend[1:]
+// Taken from: https://github.com/prometheus/prometheus/blob/f2e02c52db/storage/remote/queue_manager.go#L899-L923
+func ExtendSortedLabels(ls labels.Labels, externalLabels []labels.Label) labels.Labels {
```
Comment (MichaHoffmann) [inline pkg/store/labelpb/label.go:296]: This benchmarks slightly worse than the previous method but not terribly so: 

```
$ go test . -run XXX -bench BenchmarkExtendLabels -v
goos: linux
goarch: amd64
pkg: github.com/thanos-io/thanos/pkg/store/labelpb
cpu: 11th Gen Intel(R) Core(TM) i5-1135G7 @ 2.40GHz
BenchmarkExtendLabels
BenchmarkExtendLabels-8   	 1282364	       935.7 ns/op
PASS
ok  	github.com/thanos-io/thanos/pkg/store/labelpb	2.063s
```
vs
```
$ go test . -run XXX -bench BenchmarkExtendLabels -v
goos: linux
goarch: amd64
pkg: github.com/thanos-io/thanos/pkg/store/labelpb
cpu: 11th Gen Intel(R) Core(TM) i5-1135G7 @ 2.40GHz
BenchmarkExtendLabels
BenchmarkExtendLabels-8   	 1853936	       682.6 ns/op
PASS
ok  	github.com/thanos-io/thanos/pkg/store/labelpb	1.934s
```

File: pkg/block/block_test.go [hunk 2]
```diff
@@ -87,12 +87,12 @@ func TestUpload(t *testing.T) {
 
 	bkt := objstore.NewInMemBucket()
 	b1, err := e2eutil.CreateBlock(ctx, tmpDir, []labels.Labels{
-		{{Name: "a", Value: "1"}},
-		{{Name: "a", Value: "2"}},
-		{{Name: "a", Value: "3"}},
-		{{Name: "a", Value: "4"}},
-		{{Name: "b", Value: "1"}},
-	}, 100, 0, 1000, labels.Labels{{Name: "ext1", Value: "val1"}}, 124, metadata.NoneFunc)
+		labels.New(labels.Label{Name: "a", Value: "1"}),
```
Comment (saswatamcode) [inline pkg/block/block_test.go:90]: Wouldn't this error with stringlabels enabled?

Comment (saswatamcode) [reply pkg/block/block_test.go:90]: I think we can just use `labels.FromStrings` here

File: pkg/block/index.go [hunk 3]
```diff
@@ -673,14 +677,14 @@ func rewrite(
 			meta.Stats.NumSamples += uint64(chk.Chunk.NumSamples())
 		}
 
-		for _, l := range s.lset {
+		s.lset.Range(func(l labels.Label) {
 			valset, ok := values[l.Name]
 			if !ok {
 				valset = stringset{}
 				values[l.Name] = valset
```
Comment (saswatamcode) [inline pkg/block/index.go:684]: I think we still access `l.Name` here

**Review by saswatamcode — APPROVED** [review]
Thanks for great effort! 💪🏻 
I guess the next step will be doing this for store, and some protos.

We should probably also add some lint rules at the end.
Some questions.

File: pkg/block/block_test.go (hunk 2 above)
Comment (MichaHoffmann) [reply pkg/block/block_test.go:90]: We can but it wont error with stringlabels ( i tried actually ); so i didnt bother since its many lines to edit.

File: pkg/block/index.go (hunk 3 above)
Comment (MichaHoffmann) [reply pkg/block/index.go:684]: Sorry i dont understand! Can you rephrase please?

File: pkg/block/fetcher.go [hunk 4]
```diff
@@ -612,15 +612,16 @@ const BlockIDLabel = "__block_id"
 
 // Filter filters out blocks that have no labels after relabelling of each block external (Thanos) labels.
 func (f *LabelShardedMetaFilter) Filter(_ context.Context, metas map[ulid.ULID]*metadata.Meta, synced GaugeVec, modified GaugeVec) error {
-	var lbls labels.Labels
+	var b labels.Builder
```
Comment (yeya24) [inline pkg/block/fetcher.go:615]: I am curious if it is better to just switch to `labels.ScratchBuilder` in this case?
But maybe this is out of the scope of this pr

File: pkg/promclient/promclient.go [hunk 5]
```diff
@@ -172,26 +172,24 @@ func (c *Client) ExternalLabels(ctx context.Context, base *url.URL) (labels.Labe
...
 	var cfg struct {
 		GlobalConfig config.GlobalConfig `yaml:"global"`
 	}
 	if err := yaml.Unmarshal([]byte(d.Data.YAML), &cfg); err != nil {
-		return nil, errors.Wrapf(err, "parse Prometheus config: %v", d.Data.YAML)
+		return labels.EmptyLabels(), errors.Wrapf(err, "parse Prometheus config: %v", d.Data.YAML)
 	}
 
-	lset := cfg.GlobalConfig.ExternalLabels
-	sort.Sort(lset)
```
Comment (yeya24) [inline pkg/promclient/promclient.go:193]: No sorting here anymore? I am not sure if it is expected or we know external labels in the response will always be sorted?

File: pkg/rules/rules.go [hunk 6]
```diff
@@ -143,12 +144,9 @@ func dedupRules(rules []*rulespb.Rule, replicaLabels map[string]struct{}) []*rul
 		return rules
 	}
 
-	// Remove replica labels and sort each rule's label names such that they are comparable.
+	// Remove replica labels
 	for _, r := range rules {
 		removeReplicaLabels(r, replicaLabels)
-		sort.Slice(r.GetLabels(), func(i, j int) bool {
-			return r.GetLabels()[i].Name < r.GetLabels()[j].Name
-		})
```
Comment (yeya24) [inline pkg/rules/rules.go:151]: Is it because labels builder produces sorted labels?

File: pkg/store/labelpb/label.go [hunk 7]
```diff
@@ -292,30 +292,37 @@ func (m *ZLabel) Compare(other ZLabel) int {
...
-			ret = append(ret, lset[0])
-			lset = lset[1:]
-		} else if d > 0 {
-			ret = append(ret, extend[0])
-			extend = extend[1:]
+func ExtendSortedLabels(lset labels.Labels, extend []labels.Label) labels.Labels {
+	b := labels.NewScratchBuilder(lset.Len() + len(extend))
+	lset.Range(func(l labels.Label) {
+		for {
+			if len(extend) == 0 {
```
Comment (yeya24) [inline pkg/store/labelpb/label.go:299]: Can we have a `len(extend) == 0` check outside the function?
This should help with Cortex case where there is no external labels at all. In this case no need to go through the lset loop

File: pkg/rules/rules.go (hunk 6 above)
Comment (MichaHoffmann) [reply pkg/rules/rules.go:151]: Yeah

File: pkg/promclient/promclient.go (hunk 5 above)
Comment (MichaHoffmann) [reply pkg/promclient/promclient.go:193]: Oh for stringlabels we dont need that but if we are not building with it we probably need it still! Thanks

File: pkg/block/fetcher.go (hunk 4 above)
Comment (MichaHoffmann) [reply pkg/block/fetcher.go:615]: Yeah i would say out of scope; but generally the amount of builders probably could be .. at least pooled or something.

File: pkg/store/labelpb/label.go (hunk 7 above)
Comment (yeya24) [reply pkg/store/labelpb/label.go:299]: Oh mb... We need to copy labels anyway so probably still need to have the check.

Comment (MichaHoffmann) [reply pkg/store/labelpb/label.go:299]: I added a fast path that does `lset.Copy()` if extend is empty.

**Review by yeya24 — APPROVED** [review]
LGTM

**Review by fpetkovski — APPROVED** [review]
Do we want a CI job to verify we can compile Thanos with the stringlabels tag?

**Comment by MichaHoffmann** [comment]
> Do we want a CI job to verify we can compile Thanos with the stringlabels tag?

eventually yes, but we are not there yet!
